<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>Caveman by funpopgen</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/github-light.css">
    <script src="javascripts/scale.fix.js"></script>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1 class="header">Caveman</h1>
        <p class="header">Causal Variant Evidence Mapping using Non-parametric resampling</p>

        <ul>
          <li class="download"><a class="buttons" href="https://github.com/funpopgen/CaVEMaN/zipball/master">Download ZIP</a></li>
          <li class="download"><a class="buttons" href="https://github.com/funpopgen/CaVEMaN/tarball/master">Download TAR</a></li>
          <li><a class="buttons github" href="https://github.com/funpopgen/CaVEMaN">View On GitHub</a></li>
        </ul>

        <p class="header">This project is maintained by <a class="header name" href="https://github.com/funpopgen">funpopgen</a></p>


      </header>
      <section>
        <h1>
<a id="caveman-causal-variant-evidence-mapping-using-non-parametric-resampling" class="anchor" href="#caveman-causal-variant-evidence-mapping-using-non-parametric-resampling" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>CaVEMaN Causal Variant Evidence Mapping using Non-parametric resampling</h1>

<p>Given the results of an eQTL mapping analysis, together with the expression, genotype and phenotype files, this software will estimate the probability that the peak eQTL is causal. For more details, please see the paper (soon).</p>

<hr>

<h2>
<a id="contents" class="anchor" href="#contents" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Contents</h2>

<ol>
<li><a href="#downloading-caveman">Downloading CaVEMaN</a></li>
<li><a href="#input-data-formats">Input data formats</a></li>
<li><a href="#creating-a-single-phenotype-bed-file-from-an-eqtl-list">Creating a single phenotype bed file from an eQTL list</a></li>
<li><a href="#running-caveman">Running CaVEMaN</a></li>
<li><a href="#extracting-peak-snps-and-estimating-causal-probabilities">Extracting peak SNPs and estimating causal probabilities</a></li>
<li><a href="#building-from-source">Building from source</a></li>
</ol>

<hr>

<h2>
<a id="downloading-caveman" class="anchor" href="#downloading-caveman" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Downloading CaVEMaN</h2>

<p>CaVEMaN binaries can be downloaded from here:</p>

<p><a href="https://github.com/funpopgen/CaVEMaN/releases/tag/v1.0">https://github.com/funpopgen/CaVEMaN/releases/tag/v1.0</a></p>

<p>The standard binary is called CaVEMaN. Use CaVEMaN_CentOS if there is an error referring to GLIBC.</p>

<p>CaVEMaN requires tabix to be installed (<a href="https://github.com/samtools/htslib">https://github.com/samtools/htslib</a>).</p>

<p>If there are any issues, you can try <a href="#building-from-source">building from source</a>, or contact <a href="mailto:andrew.brown@unige.ch">andrew.brown@unige.ch</a>.</p>

<hr>

<h2>
<a id="input-data-formats" class="anchor" href="#input-data-formats" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Input data formats</h2>

<p>CaVEMaN uses the same data formats as fastQTL (<a href="http://fastqtl.sourceforge.net/">http://fastqtl.sourceforge.net/</a>), which the exception that the bed file should be uncompressed. To convert the fastQTL bed file run the following line:</p>

<pre><code>gunzip myphenotypefile.bed.gz
</code></pre>

<p>The bed file should only contain genes for which an eQTL has been discovered, unless a set of conditional eQTLs have been produced, in which case the next section explains how to produce a new bed file from this list. This requires a list of eQTL, which should be a tab-separated file with no header, and the following fields to uniquely determine the SNP-gene pairs:</p>

<ol>
<li>Gene</li>
<li>Chromosome</li>
<li>Base pair location</li>
<li>Reference allele</li>
<li>Alternate allele</li>
</ol>

<p>The --correct option can also take a covariates file. This should be tab separated, the first row should contain subject IDs, and there should be one row for each covariate.</p>

<hr>

<h2>
<a id="creating-a-single-phenotype-bed-file-from-an-eqtl-list" class="anchor" href="#creating-a-single-phenotype-bed-file-from-an-eqtl-list" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Creating a single phenotype bed file from an eQTL list</h2>

<p>CaVEMaN is based on the assumption that there is only one eQTL signal in the cis window. In cases where multiple eQTLs are known for a given gene, we can produce "single signal" phenotypes by regressing out all but one eQTL at a time (and optionally covariates as well). Given an expression bed file (expression.bed), a genotype file (genotype.vcf.gz) and a list of eQTL (eqtl.list), we produce the file with the following command:</p>

<pre><code> CaVEMaN --correct eqtl.list --bed expression.bed --vcf genotype.vcf.gz \
     --out corrected.expression.bed
</code></pre>

<p>If a set of important covariates are known (covariates.txt) were included in the eQTL mapping, these can be considered as well:</p>

<pre><code> CaVEMaN --correct eqtl.list --bed expression.bed --vcf genotype.vcf.gz \
     --out corrected.expression.bed --cov covariates.txt
</code></pre>

<hr>

<h2>
<a id="running-caveman" class="anchor" href="#running-caveman" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Running CaVEMaN</h2>

<p>To run CaVEMaN, all that is necessary is a bed file containing expression values (expression.bed) and a vcf file containing genotype values (genotype.vcf.gz, either the DS or GT field must be present).</p>

<p>CaVEMaN has been designed so that the full analysis can be broken up into chunks, which can be submitted as cluster jobs separately. This requires two flags (if neither are present the whole analysis will be submitted), --genes specifies the number of genes to be analysed in each job, --job-number indexes the job. Therefore, if --genes 10 is specified, --job-number 1 will analyse genes 1-10 in the bed file, --job-number 2 will process genes 11-20 and so on.</p>

<p>To submit a job array, with a bed file of 9,995 genes, where each job analyses 50 genes (9,995 / 50 = 199.9, so we need to run 200 jobs to cover all genes), we would submit the following command to an LSF cluster:</p>

<pre><code> bsub -o out -J"CaVEMaN[1-200]" \
     "CaVEMaN --bed expression.bed --vcf genotype.vcf.gz --genes 50 \
     --job-number \$LSB_JOBINDEX --out results\$LSB_JOBINDEX"
</code></pre>

<p>In addition, the --perm option allows you to set the number of bootstrap samples used in the analysis.</p>

<p>This command will produce 200 results files, results1-results200. To concatenate them together run:</p>

<pre><code> awk 'FNR&gt;1||NR==1' results* &gt; results.all
</code></pre>

<p>The output should look like this:</p>

<p><img src="screenshots/CaVEMaN.results.png" alt="CaVEMaN results example"></p>

<p>The output contains every SNP in the cis window, the first five fields specify the gene and SNP, Cor and P fields give the results of a simple correlation analysis of SNP and gene, and CaVEMaN gives the CaVEMaN score, a measure of whether this SNP is causal.</p>

<hr>

<h2>
<a id="extracting-peak-snps-and-estimating-causal-probabilities" class="anchor" href="#extracting-peak-snps-and-estimating-causal-probabilities" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Extracting peak SNPs and estimating causal probabilities</h2>

<p>Once the CaVEMaN analysis is complete, we can parse the results file to extract the SNP most likely to be causal, and calculate its causal probability (using the results of simulations). To do this run the following command:</p>

<pre><code> CaVEMaN --best results.all --out results.best
</code></pre>

<p>The file results.best should look like this:</p>

<p><img src="screenshots/CaVEMaN.best.png" alt="Screenshot, peak eQTL"></p>

<p>The file contains the same fields as before, with the addition of the "Probability" column which gives the probability the SNP is causal.</p>

<hr>

<h2>
<a id="building-from-source" class="anchor" href="#building-from-source" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Building from source</h2>

<p>This requires a D compiler and a version of the GNU scientific library, version 2.1 or above.</p>

<h3>
<a id="cloning-the-repository" class="anchor" href="#cloning-the-repository" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Cloning the repository</h3>

<p>First, clone the repository by running:</p>

<pre><code>git clone https://github.com/funpopgen/CaVEMaN.git
</code></pre>

<p>This should create a folder called CaVEMaN. Inside this folder is a file called makefile, in the first four lines we can specify the location of the D compiler and also where the gsl library is (if it is not already installed on your cluster).</p>

<h3>
<a id="download-a-d-compiler" class="anchor" href="#download-a-d-compiler" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Download a D compiler</h3>

<p>Then, download a D compiler from either here: <a href="https://github.com/ldc-developers/ldc/releases">https://github.com/ldc-developers/ldc/releases</a> or here: <a href="http://dlang.org/dmd-linux.html">http://dlang.org/dmd-linux.html</a> (ldc produces faster software, my experience is that dmd works better on older operating systems). Decompress the downloaded file, then edit either the third or fourth line of the makefile in the CaVEMaN folder so that it contains the full path to the relevant compiler, i.e. either:</p>

<p><strong>DMD = /path/to/dmd2/linux/bin64/dmd</strong></p>

<p>or</p>

<p><strong>LDC = /path/to/ldc2-1.1.0-beta2-linux-x86_64/bin/ldc2</strong></p>

<h3>
<a id="installing-if-necessary-gsl-v21" class="anchor" href="#installing-if-necessary-gsl-v21" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Installing, if necessary, gsl v2.1</h3>

<p>First, check if a sufficiently recent version of the gsl is already installed by running:</p>

<pre><code> gsl-config --version
</code></pre>

<p>If this states that version 2.1 or more recent is installed, you can move on to building the software.</p>

<p>Otherwise, download the gsl from here: <a href="http://mirror.switch.ch/ftp/mirror/gnu/gsl/gsl-2.1.tar.gz">http://mirror.switch.ch/ftp/mirror/gnu/gsl/gsl-2.1.tar.gz</a></p>

<p>Create a folder to install it to called gsl, then run the following commands:</p>

<pre><code>tar -xf gsl-2.1.tar.gz
cd gsl-2.1/
./configure prefix=/path/to/gsl
make
make install
</code></pre>

<p>Then, edit the second line of the makefile in the CaVEMaN folder to give it the full path to the gsl folder you created, i.e.</p>

<p><strong>GSL = /path/to/gsl</strong></p>

<h3>
<a id="building-the-software" class="anchor" href="#building-the-software" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Building the software</h3>

<p>Now to compile, run the commands from the relevant cell in the table (which depend on which compiler you downloaded and whether the gsl was pre-installed).</p>

<table>
<thead>
<tr>
<th></th>
<th>ldc</th>
<th>dmd</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>GSL already installed</strong></td>
<td>make test &amp;&amp; make</td>
<td>make dmd_test &amp;&amp; make dmd</td>
</tr>
<tr>
<td><strong>GSL installed by  you</strong></td>
<td>make static_test &amp;&amp; make static</td>
<td>make dmd_static_test &amp;&amp; make dmd_static</td>
</tr>
</tbody>
</table>

<p>Now the CaVEMaN binary should be within the bin folder in the CaVEMaN directory, running</p>

<pre><code> ./bin/CaVEMaN --help
</code></pre>

<p>should bring up the help.</p>
      </section>
      <footer>
        <p><small>Hosted on <a href="https://pages.github.com">GitHub Pages</a> using the Dinky theme</small></p>
      </footer>
    </div>
    <!--[if !IE]><script>fixScale(document);</script><![endif]-->
		
  </body>
</html>
